### PR TITLE
core/state: parallelise parts of state commit

### PR DESCRIPTION
**Pull request by karalabe**
This PR attempts to run bits and pieces of statedb.Commit concurrently to one another. The observation is that even though the code is very light, concurrency does seem to make a difference. Will post some benchmark results. Edit:

Screenshot 2024-04-29 at 20 39 43

Also, I think it's possible to make the account update run concurrently with the storage updates, which might shave even more off. Will need to run that as a benchmark too. Edit:

Screenshot 2024-04-30 at 09 07 31

### CODE REVIEW

File: internal/workerpool/workerpool.go
Comment (fjl) [inline internal/workerpool/workerpool.go:18]: I honestly feel this could be in `internal/syncx`, which is our existing package for synchronization primitives.

Comment (karalabe) [reply internal/workerpool/workerpool.go:18]: hah lol, I didn't see that package. will move

File: internal/syncx/workerpool.go
Comment (fjl) [reply internal/syncx/workerpool.go:35]: Hmm, but actually, since your `R` is just `error`, the `WorkerPool` then becomes equivalent to https://pkg.go.dev/golang.org/x/sync/errgroup with a limit, so you could just as well use that:

```go
var pool errgroup.Group
pool.SetLimit(threads)
pool.Go(func () error {
    // commit state object
})
...
pool.Wait()
```

Comment (fjl) [reply internal/syncx/workerpool.go:35]: `errgroup.Group` might actually be better, since it also cancels all work on the first error.

Comment (karalabe) [reply internal/syncx/workerpool.go:35]: That catch is that that will promote just dumping everything in there, mixing tasks that have different resource bounds. I can. imagine that simple clogging the whole thing because one resource holds off another type.

E.g If I merge CPU and disk tasks, then one thread could block on disk, not doing any CPU processing any more because of it. With more typed pools, it's harder to mix and match (and uglier API to make more than one thing concurrent). Hence my thought with the pipeline...

Basically what we want is CPU to be running concurrently on all cores, but not more to not trash. On the other hand, we want to run disk as parallel as possible, since we're going to be more limited thread wise than the SSD's capacity anyway. So that's a different concurrency than the pool.

Comment (karalabe) [reply internal/syncx/workerpool.go:35]: errgroup is different, that blocks submitting a task when all the threads are busy. also errgroup is a new goroutine for each task vs. this one reusing goroutines to make the creation overhead smaller for very tiny tasks

**Review by Talhaarain150 — APPROVED** [review]

File: internal/syncx/workerpool.go
Comment (karalabe) [reply internal/syncx/workerpool.go]: Benchmarked a lot. It was very inconclusive whethre workerpool or errgroup is worth it. So I went with errgroup, since it's something we don't have to maintain. for now at least

File: core/state/statedb.go
Comment (holiman) [inline core/state/statedb.go:1193]: This looks wrong. Or is the idea that the last finisher is the one that sets it?

Comment (karalabe) [reply core/state/statedb.go:1193]: There's only one account commit, ins't it? Even in pre-Byzantium, we only call one commit per account trie in a block

Comment (holiman) [reply core/state/statedb.go:1193]: Right, this is not a loop

**Review by rjl493456442 — APPROVED** [review]

File: core/state/statedb.go
Comment (fjl) [inline core/state/statedb.go:1224]: Just a quick note, this locking scheme here will definitely affect performance. You're spawning concurrent work but it'll all have to synchronize after. It would be better to pre-allocate a slice of suitable size to track the output sets (or even have it as a field in StateObject), then perform the merge operation and counter updates after all commits are done.

Comment (karalabe) [reply core/state/statedb.go:1224]: If I were to wait until all results are in and then merge, then the merge wound't proceed concurrnetly with the disk reads. IMO that would be worse.

I could make a channel and stream the results out and have an outer goroutine listen and merge one by one, but that would just end up being the same as locking and allowing access one by one via the mutex.

Don't really see why it would be different.

Comment (fjl) [reply core/state/statedb.go:1224]: The channel way would also work, and the channel could even be buffered. Writes on buffered channels are fast even under contention AFAIK, so could be an idea.

**Pull request merged**
